pkg/suggestion/v1alpha3/goptuna/sample.go

```diff
@@ -0,0 +1,153 @@
+				Name:  name,
+				Value: strconv.FormatFloat(p, 'f', -1, 64),
+			})
+		case goptuna.IntUniformDistribution:
+			p, err := trial.SuggestInt(name, distribution.Low, distribution.High)
+			if err != nil {
+				klog.Errorf("Failed to get suggested param: %s", err)
+				return nil, err
+			}
+			assignments = append(assignments, &api_v1_alpha3.ParameterAssignment{
+				Name:  name,
+				Value: strconv.Itoa(p),
+			})
+		case goptuna.StepIntUniformDistribution:
+			p, err := trial.SuggestStepInt(name, distribution.Low, distribution.High, distribution.Step)
+			if err != nil {
+				klog.Errorf("Failed to get suggested param: %s", err)
+				return nil, err
+			}
+			assignments = append(assignments, &api_v1_alpha3.ParameterAssignment{
+				Name:  name,
+				Value: strconv.Itoa(p),
+			})
+		case goptuna.CategoricalDistribution:
+			p, err := trial.SuggestCategorical(name, distribution.Choices)
+			if err != nil {
+				klog.Errorf("Failed to get suggested param: %s", err)
+				return nil, err
+			}
+			assignments = append(assignments, &api_v1_alpha3.ParameterAssignment{
+				Name:  name,
+				Value: p,
+			})
+		}
+	}
+
+	if t, err := study.Storage.GetTrial(trial.ID); err == nil {
+		klog.Infof("Success to sample new trial: trialID=%d, params=%v", t.ID, t.Params)
+	}
+	return assignments, nil
+}
+
+func findGoptunaTrialIDByParam(study *goptuna.Study, trialMapping map[string]int, ktrial goptuna.FrozenTrial) (int, error) {
+	trials, err := study.GetTrials()
+	if err != nil {
+		return -1, err
+	}
+
+	existInMapping := func(trialID int) bool {
+		for j := range trialMapping {
+			if trialMapping[j] == trialID {
+				return true
+			}
+		}
+		return false
+	}
+
+	minManhattan := math.MaxFloat64
+	estimatedTrialID := -1
+	for i := range trials {
+		if trials[i].State != goptuna.TrialStateRunning {
+			continue
+		}
+
+		// skip the trial id which already mapped by other katib trial name
+		if existInMapping(trials[i].ID) {
+			continue
+		}
+
+		// To understand how this function estimate the Goptuna trial ID from the parameters,
+		// you need to understand the 'internal representation' in Goptuna.
+		// Goptuna trials holds the parameters in two types of representations.
+		// To explain the representation format, please see the following example.
+		//
+		// * Search space: map[string]interface{}{"x1": Uniform{Min: -10, Max: 10}, "x2": Categorical{Choices: []string{"param-1", "param-2", "param-3"}}}
+		// * External representation: map[string]interface{}{"x1": 5.5, "x2": "param-2"}
+		// * Internal representation: map[string]float64{"x1": 5.5, "x2": 1.0}
+		//
+		// In the internal representation, all parameters are represented by `float64` to store the storage
+		// (because Goptuna supports not only in-memory but also RDB storage backend).
+		// To represent categorical parameters, Goptuna holds an index of the list in the database.
+		//
+		// This function calculates Manhattan distance of internal representation parameters.
```

Comment:
Since `gTrial` generates `kTrial` we can't just check that `gtrialParamValue = ktrialParamValue`?
Why we calculate Manhattan between parameters and look for "similar" parameters?
Did I misunderstand something here?

Reply:
I'm just not confident that the parameter value actually kept the original precision. But on second thought, all parameter values are converted to string before sending `katib-controller`. If `katib-controller` holds it as just a string, It seems no problems to just compare exactly as you said.

I'll check the behavior of `gtrialParamValue = ktrialParamValue`. Thank you for your review!

Reply:
It seems to work fine 🎉   I've just pushed it. Thank you for your comment!

```
$ kube-prompt
>>> logs -f example-1-cmaes-f84789cc6-cx8g7 | grep mapping
I0413 14:56:34.614940       1 service.go:97] Update trial mapping : trialName=example-1-p2g98q46 -> trialID=1
I0413 14:56:34.614959       1 service.go:97] Update trial mapping : trialName=example-1-d572g657 -> trialID=0
I0413 14:56:35.828050       1 service.go:97] Update trial mapping : trialName=example-1-fk48npm5 -> trialID=2
I0413 14:56:43.749274       1 service.go:97] Update trial mapping : trialName=example-1-k4l5lwtq -> trialID=3
I0413 14:56:46.481446       1 service.go:97] Update trial mapping : trialName=example-1-xrbtv2p4 -> trialID=4
I0413 14:56:54.147858       1 service.go:97] Update trial mapping : trialName=example-1-zs2gdpj8 -> trialID=5
I0413 14:56:56.936444       1 service.go:97] Update trial mapping : trialName=example-1-vp6dznkl -> trialID=6
I0413 14:57:03.774500       1 service.go:97] Update trial mapping : trialName=example-1-bhb78f6n -> trialID=7
I0413 14:57:06.349886       1 service.go:97] Update trial mapping : trialName=example-1-t5477bm5 -> trialID=8
I0413 14:57:13.965154       1 service.go:97] Update trial mapping : trialName=example-1-89bm85cc -> trialID=9
I0413 14:57:16.335771       1 service.go:97] Update trial mapping : trialName=example-1-v96cktcg -> trialID=10
I0413 14:57:25.184541       1 service.go:97] Update trial mapping : trialName=example-1-62p8dx96 -> trialID=11
I0413 14:57:28.127907       1 service.go:97] Update trial mapping : trialName=example-1-hq9r44zn -> trialID=12
I0413 14:57:35.801153       1 service.go:97] Update trial mapping : trialName=example-1-c7dlrdt9 -> trialID=13
I0413 14:57:37.414098       1 service.go:97] Update trial mapping : trialName=example-1-tcpm8224 -> trialID=14
I0413 14:57:46.445595       1 service.go:97] Update trial mapping : trialName=example-1-6qfpd7gm -> trialID=15
I0413 14:57:48.662856       1 service.go:97] Update trial mapping : trialName=example-1-z6t6wqb6 -> trialID=16
I0413 14:57:56.507626       1 service.go:97] Update trial mapping : trialName=example-1-q7ck6sq2 -> trialID=17
I0413 14:57:58.449231       1 service.go:97] Update trial mapping : trialName=example-1-g24xvj6j -> trialID=18
I0413 14:58:06.930626       1 service.go:97] Update trial mapping : trialName=example-1-gctzc2v6 -> trialID=19
I0413 14:58:08.043766       1 service.go:97] Update trial mapping : trialName=example-1-qjn488pt -> trialID=20
I0413 14:58:16.141586       1 service.go:97] Update trial mapping : trialName=example-1-76q2r8gr -> trialID=21
I0413 14:58:19.100734       1 service.go:97] Update trial mapping : trialName=example-1-hmmkcd5v -> trialID=22
I0413 14:58:25.959862       1 service.go:97] Update trial mapping : trialName=example-1-6g4nf7wn -> trialID=23
I0413 14:58:29.313208       1 service.go:97] Update trial mapping : trialName=example-1-pl6l829x -> trialID=24
I0413 14:58:37.596133       1 service.go:97] Update trial mapping : trialName=example-1-pbnqntxp -> trialID=25
I0413 14:58:39.360219       1 service.go:97] Update trial mapping : trialName=example-1-5dhshjll -> trialID=26
I0413 14:58:47.193674       1 service.go:97] Update trial mapping : trialName=example-1-jfprnznc -> trialID=27
I0413 14:58:50.347170       1 service.go:97] Update trial mapping : trialName=example-1-h8tnxvb5 -> trialID=28
I0413 14:58:57.282617       1 service.go:97] Update trial mapping : trialName=example-1-fq2699q8 -> trialID=29
I0413 14:58:59.922027       1 service.go:97] Update trial mapping : trialName=example-1-gtrqtlcl -> trialID=30
I0413 14:59:07.513689       1 service.go:97] Update trial mapping : trialName=example-1-wg9rtkkm -> trialID=31
I0413 14:59:10.021879       1 service.go:97] Update trial mapping : trialName=example-1-qhbzkzld -> trialID=32
I0413 14:59:17.553238       1 service.go:97] Update trial mapping : trialName=example-1-v4fn2ml4 -> trialID=33
I0413 14:59:20.678341       1 service.go:97] Update trial mapping : trialName=example-1-66xhfm9z -> trialID=34
I0413 14:59:28.029968       1 service.go:97] Update trial mapping : trialName=example-1-g2pwjqpb -> trialID=35
I0413 14:59:30.981466       1 service.go:97] Update trial mapping : trialName=example-1-nhqbgvlv -> trialID=36
I0413 14:59:39.416874       1 service.go:97] Update trial mapping : trialName=example-1-xbq757hw -> trialID=37
I0413 14:59:42.396212       1 service.go:97] Update trial mapping : trialName=example-1-hr2c229v -> trialID=38
I0413 14:59:50.084652       1 service.go:97] Update trial mapping : trialName=example-1-6sgkx52b -> trialID=39
I0413 14:59:51.725609       1 service.go:97] Update trial mapping : trialName=example-1-gclf8phc -> trialID=40
:
```